src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala

```diff
@@ -176,21 +178,46 @@ object RuleUtils {
     //        Project(A,B) -> Filter(C = 10) -> Scan (A,B,C,D,E)
     // in the presence of a suitable index, the getIndexPlan() method will emit:
     //        Project(A,B) -> Filter(C = 10) -> Index Scan (A,B,C)
-    plan transformDown {
+    plan transformUp {
       case baseRelation @ LogicalRelation(_: HadoopFsRelation, baseOutput, _, _) =>
         val location =
           new InMemoryFileIndex(spark, index.content.files, Map(), None)
+
+        val updatedSchema = if (index.excludedFiles.isEmpty) {
+          StructType(index.schema.filter(baseRelation.schema.contains))
+        } else {
+          // Include DATA_FILE_NAME_COLUMN column in schema.
+          assert(index.hasLineageColumn(spark))
+          index.schema
+        }
+
         val relation = HadoopFsRelation(
           location,
           new StructType(),
-          StructType(index.schema.filter(baseRelation.schema.contains(_))),
+          updatedSchema,
           if (useBucketSpec) Some(index.bucketSpec) else None,
           new ParquetFileFormat,
           Map())(spark)
 
         val updatedOutput =
           baseOutput.filter(attr => relation.schema.fieldNames.contains(attr.name))
-        baseRelation.copy(relation = relation, output = updatedOutput)
+
+        if (index.excludedFiles.isEmpty) {
+          baseRelation.copy(relation = relation, output = updatedOutput)
+        } else {
+          // Exclude deleted index entries coming from source data files listed as "excluded".
+          // This is done by transforming plan from a single Scan node on index files to
+          // "Scan -> Filter -> Project" in which:
+          // - Scan's output schema lists all index columns which includes lineage column.
+          // - Filter's condition uses lineage column to exclude deleted index entries.
+          // - Project fixes output schema by removing lineage column from index scan results.
+          val lAttr = AttributeReference(IndexConstants.DATA_FILE_NAME_COLUMN, StringType)(
+            NamedExpression.newExprId)
+          val deletedFileNames = index.excludedFiles.map(Literal(_)).toArray
+          val rel = baseRelation.copy(relation = relation, output = updatedOutput ++ Seq(lAttr))
+          val filter = Filter(condition = Not(In(lAttr, deletedFileNames)), rel)
```

Comment:
@pirz One thing to evaluate here is what happens when a large number of files are deleted. For instance, I can imagine the query plan can get pretty big. Can you run some quick benchmark to test this for a few 1000 files deleted? I think this will have implication on query compilation time. An alternate way of doing this would be to express this as a `JOIN` against another table containing the list of deleted files. 

CC: @imback82 @sezruby @apoorvedave1 for a second opinion in case they think this is unnecessary.

Reply:
It would be good if @pirz could run TPCH with 100k chunks dataset for the injected filter condition of deleted files while I'm OOF this week :)

0. build index with lineage column on full dataset (lineitem/part 0~99999 files)
1. remove files
    - case 1) delete 200 files
    - case 2) delete 5000 files
    - case 3) delete 25k files
    - case 4) delete 50k files
2. run
   - run w/o index
   - run w/o refreshed index => similar to partial index
   - run w/ refreshed index => similar to hybrid scan, delete-only
   - run w/ fully refreshed index 
 
I usually run 4x4 cases but I think it's better to test (5k or 25k deleted files with refreshed index) quickly to see the regression.
(Jfyi, deleting files on remote storage took quite a lot of time from my experience...)